src/manifold/deferred.clj

```diff
@@ -1386,6 +1512,182 @@
 
 (prefer-method print-method IDeferred IDeref)
 
+
+(defmacro ^:no-doc def-async-for
+  "Defines a CompletionStage async version of the function associated with
+   the given symbol."
+  [fn-name]
+  (let [async-name (symbol (str (name fn-name) "-async"))]
+    `(defn ~async-name
+       ([d# f#]
+        (~async-name d# f# (or (ex/executor) (ex/execute-pool))))
+       ([d# f# executor#]
+        (~fn-name (onto d# executor#) f#)))))
+
+(defmacro ^:no-doc def-async-for-dual
+  "Defines a CompletionStage async version of the two-deferred
+   function associated with the given symbol."
+  [fn-name]
+  (let [async-name (symbol (str (name fn-name) "-async"))]
+    `(defn ~async-name
+       ([d# d2# f#]
+        (~async-name d# d2# f# (or (ex/executor) (ex/execute-pool))))
+       ([d# d2# f# executor#]
+        (~fn-name (onto d# executor#) d2# f#)))))
+
+(defmacro ^:no-doc assert-some
+  "Throws NullPointerException if any of the arguments is null."
+  [& values]
+  `(do ~@(for [value values]
+           `(let [value# ~value]
+              (when (nil? value#)
+                (throw (NullPointerException. ~(str value " was null"))))))))
+
+(defn- map-deferred
+  "Like map/fmap but for deferreds.
+
+   This function does not unwrap the result of f"
+  [d f]
+  (let [d' (deferred)]
+    (on-realized d
+                 (fn [val] (success! d' (f val)))
+                 (fn [error] (error! d' error)))
+    d'))
+
+(defn- mapcat-deferred
```

Comment:
These names are a bit confusing in Clojure. I assume they're Haskell-y? `map` and `mapcat` in Clojure are exclusively for sequences. It doesn't quite make sense to speak of using them for wrapping objects.

I don't mind calling `map-deferred` fmap if that's more accurate. Not sure what else to call `mapcat-deferred`.

Reply:
`mapcat-deferred` is only used in one place. Let's just inline its body and remove it.

Reply:
No problem. Despite the inlining circunstance, the `haskell`ier name for `mapcat-deferred` here would probably be `bind-deferred`.